operator/pkg/ciliumendpointbatch/endpointbatch.go

```diff
@@ -0,0 +1,300 @@
+		ciliumEndpointBatchStore:            cebStore,
+		ciliumEndpointBatchUpsertSyncPeriod: cebSyncPeriod,
+		ciliumEndpointBatchDeleteSyncPeriod: cebDeleteSyncPeriod,
+	}
+}
+
+// start the worker thread, reconciles the modified CEBs with api-server
+func (c *CiliumEndpointBatchController) Run(ces cache.Indexer, stopCh chan struct{}) {
+
+	log.Info("Bootstrap ceb controller")
+	defer utilruntime.HandleCrash()
+	defer c.queue.ShutDown()
+
+	// Cache CiliumEndpointStore Interface locally
+	c.ciliumEndpointStore = ces
+
+	// On operator warm boot, remove stale cep entries present in CEB
+	c.ciliumRemoveStaleCepEntries()
+
+	log.WithFields(logrus.Fields{
+		"Upsert syncperiod": c.ciliumEndpointBatchUpsertSyncPeriod,
+		"Delete syncperiod": c.ciliumEndpointBatchDeleteSyncPeriod,
+	}).Debug("Starting CEB controller reconciler.")
+
+	// TODO: multiple worker threads can run concurrently to reconcile with api-server
+	go wait.Until(c.worker, c.ciliumEndpointBatchUpsertSyncPeriod, stopCh)
+
+	go func() {
+		defer utilruntime.HandleCrash()
+	}()
+
+	<-stopCh
+
+	return
+}
+
+// Upon warm boot[restart], Iterate over all CEPs which we got from the api-server
+// and compare it with CEPs packed inside CEB.
+// If there are any stale CEPs present in CEBs, remove them from their CEB.
+func (c *CiliumEndpointBatchController) ciliumRemoveStaleCepEntries() {
+	log.Info("Remove stale CEP entries in CEB")
+
+	// Get all ceps from local datastore
+	staleCeps := c.Manager.getAllCepNames()
+
+	// Remove stale CEP entries present in CEB
+	for _, cepName := range staleCeps {
+		if _, exists, err := c.ciliumEndpointStore.GetByKey(cepName); err == nil && exists || err != nil {
+			continue
+		}
+		log.WithFields(logrus.Fields{
+			"cep-name": cepName,
+		}).Debug("Removing stale CEP entry.")
+		c.Manager.RemoveCepFromCache(cepName)
+	}
+	return
+}
+
+// Sync all cebs from cebStore to manager cache.
+// Note: CEBs are synced locally before CEB controller running and this is required.
+func syncCebsInLocalCache(cebStore cache.Store, manager cebManager) {
+
+	for _, obj := range cebStore.List() {
+		ceb := obj.(*v2alpha1.CiliumEndpointBatch)
+		// If CEB is already cached locally, do nothing.
+		if _, err := manager.getCebFromCache(ceb.GetName()); err == nil {
+			continue
+		}
+
+		// Create new CEB locally, with the given cebName
+		manager.createCeb(ceb.GetName())
+
+		// Deep copy the ceb, we got from api-server to local datastore.
+		manager.updateCebInCache(ceb, true)
+
+	}
+	log.Debug("Successfully synced all CEBs locally")
+	return
+}
+
+// worker runs a worker thread that just dequeues items, processes them, and
+// marks them done. You may run as many of these in parallel as you wish; the
+// workqueue guarantees that they will not end up processing the same ceb
+// at the same time
+func (c *CiliumEndpointBatchController) worker() {
+	for c.processNextWorkItem() {
+	}
+}
+
+func (c *CiliumEndpointBatchController) processNextWorkItem() bool {
```

Comment:
nit: needs a comment for this function

Reply:
this is pretty naive function for all k8s controllers :)